man/sigconvert.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "SIGCONVERT" "1" "10 June 2026" "" "The Regina Handbook"

.SH NAME
sigconvert \- Convert between generations of isomorphism signature or    knot/link signature
.SH SYNOPSIS

\fBsigconvert\fR [ \fB-k, --knot\fR | \fB-2, --dim2\fR | \fB-3, --dim3\fR | \fB-4, --dim4\fR | \fB-d, --dim=\fIdimension\fB\fR ] [ \fB-g, --gen=\fIgeneration\fB\fR ] [ \fB--\fR ] { \fB\fIsignature\fB\fR\fI ...\fR | \fB-\fR }


\fBsigconvert\fR { \fB-v, --version\fR | \fB-?, --help\fR }

.SH "DESCRIPTION"
.PP
Converts between first-generation and second-generation isomorphism
signatures for triangulations, or between first-generation and
second-generation knot/link signatures.
.PP
In Regina, triangulations and knot/link diagrams are often represented
using \fIisomorphism signatures\fR and
\fIknot/link signatures\fR respectively.
These signatures are short pieces of text that identify a triangulation
or link diagram uniquely up to combinatorial isomorphism (which for
triangulations includes relabellings of top-dimensional simplices and
their vertices, and for knot/link diagrams includes relabellings of
crossings as well as various rotation/reflection/reversal operations).
.PP
The format of these signatures changed in Regina\~8.0.
The old signatures, used in Regina\~<=\~7.x, are called
\fIfirst-generation\fR signatures.
The new signatures, introduced in Regina\~8.0, are called
\fIsecond-generation\fR signatures.
Second-generation signatures are smaller, often faster to compute, and are
recommended for use in new projects.
Nowadays you should only use first-generation signatures if you need to
maintain compatibility with older versions of Regina.
.PP
The purpose of \fBsigconvert\fR is simply to convert
between these two formats.  It will take a list of signatures as
input, convert them to the requested format (second-generation by default,
or first-generation if you pass \fB--gen=1\fR), and write the
converted signatures to standard output, each on a separate line.
.PP
The input signatures are assumed to represent 3-dimensional triangulations
unless one of the other link or dimension options is passed
(\fB--knot\fR, \fB--dim2\fR,
\fB--dim4\fR, or \fB--dim=...\fR).
Your input may include first-generation signatures, second-generation
signatures, or a mix of both.
.PP
Note that, although second-generation signatures can optionally encode
orientation, this tool will always output \fBunoriented\fR
signatures (since first-generation signatures did not encode orientation
at all).  A side-effect is that, if you need to "unorient" a
second-generation signature, you can simply run it through this tool
(as shown in the last example below).
.PP
There are two ways to provide the input signatures.
The simplest way is to pass them as command-line arguments.
Alternatively (for example, if you have a large number of signatures to
convert), you can pass a dash (\fB-\fR)
on the command-line instead, which will instruct
\fBsigconvert\fR to read the
input signatures from standard input (one per line).
Both methods are illustrated in the examples below.
.sp
.RS
.B "Tip:"
Very large triangulations or links have signatures that begin with a dash
(-).  Such signatures could be mistaken for options
when passing them on the command line.  To avoid this, you can pass the
special option \fB--\fR immediately before the signatures:
this indicates that all command-line options have finished,
and whatever comes next should be treated as input signatures.

.nf
    example$ \fBsigconvert -- -b-LLvALwvM...\fR
.fi
.RE
.SH "OPTIONS"
.TP
\fB-k, --knot\fR
Indicates that the input signatures represent knots or links.
.TP
\fB-2, --dim2\fR
Indicates that the input signatures represent 2-dimensional
triangulations.
.TP
\fB-3, --dim3 (default)\fR
Indicates that the input signatures represent 3-dimensional
triangulations.

This is the default if none of the arguments
\fB--knot\fR,
\fB--dim2\fR, \fB--dim3\fR,
\fB--dim4\fR,
or \fB--dim=...\fR are passed.
.TP
\fB-4, --dim4\fR
Indicates that the input signatures represent 4-dimensional
triangulations.
.TP
\fB-d, --dim=\fIdimension\fB\fR
Indicates that the input signatures represent triangulations of the
given dimension.  Here you can pass any dimension supported by
Regina (including higher dimensions >=\~5).

If none of the arguments \fB--knot\fR,
\fB--dim2\fR, \fB--dim3\fR,
\fB--dim4\fR or \fB--dim=...\fR
are passed, then the default is to assume triangulations of
dimension\~3.
.TP
\fB-g, --gen=\fIgeneration\fB\fR
Indicates whether to output first-generation signatures or
second-generation signatures.

The default is to output second-generation signatures.
.TP
\fB--\fR
Indicates that all other options have finished, and whatever comes
next on the command line should be treated as the input signatures.

This is useful when you have a signature that begins with a dash,
to avoid confusing your input signature with a regular command line
option.
.TP
\fB-v, --version\fR
Show which version of Regina is being used, and exit
immediately.
.TP
\fB-?, --help\fR
Display brief usage information, and exit immediately.
.SH "EXAMPLES"
.PP
The following example converts back and forth between 3-dimensional
isomorphism signature formats for the figure eight knot complement.
Here the first-generation signature is
cPcbbbiht, and the second-generation
signature is cV6cqb:

.nf
    example$ \fBsigconvert cPcbbbiht\fR
    cV6cqb
    example$ \fBsigconvert -g1 cV6cqb\fR
    cPcbbbiht
    example$
.fi
.PP
The next example converts knot/link signatures from first-generation
to second-generation, for both the Whitehead link and the Borromean rings:

.nf
    example$ \fBsigconvert -k fabcadefbcedvfpd gabcdgaecfgbfdeLwto\fR
    f3q7Gko
    gVc1JwUJ
    example$
.fi
.PP
Suppose you have a file \fIsigs.txt\fR containing
first-generation 3-dimensional isomorphism signatures (one per line).
For instance, you might have obtained this from running
tricensus --sigs using an older version of Regina.
To convert these into second-generation signatures, you can instruct
\fBsigconvert\fR to use standard input:

.nf
    example$ \fBcat sigs.txt\fR
    eLAkbcbddhhwhr
    eLAkbcbddhhwqj
    eLAkbcbddhxwqs
    \&...
    example$ \fBcat sigs.txt | sigconvert -\fR
    epAA3aaGi
    ehrBRr4cl
    ehrBRrici
    \&...
    example$
.fi
.PP
Our last example concerns orientation.  As discussed above,
second-generation signatures can optionally encode orientation; however,
\fBsigconvert\fR will always output unoriented signatures.
The example below illustrates: (i)\~creating two lens spaces in
Python with opposite orientations, (ii)\~computing their oriented
signatures (bNjc and bNld),
and then (iii)\~using \fBsigconvert\fR to
"unorient" the signatures (whereupon the two signatures become
identical).

.nf
    example$ \fBregina-python\fR
    Regina 8.0
    Software for low-dimensional topology
    Copyright (c) 1999-2026, The Regina development team
    >>> x = Example3.lens(4,1)
    >>> y = Example3.lens(4,1)
    >>> y.reflect()
    >>> print(x.neoSig(True)) # True -> oriented signature
    bNjc
    >>> print(y.neoSig(True)) # True -> oriented signature
    bNld
    >>>
    example$ \fBsigconvert bNjc bNld\fR
    bNjc
    bNjc
    example$
.fi
.SH "MACOS USERS"
.PP
If you downloaded a drag-and-drop app bundle, this utility is
shipped inside it.  If you dragged Regina to the main
Applications folder, you can run it as
/Applications/Regina.app/Contents/MacOS/sigconvert\&.
.SH "WINDOWS USERS"
.PP
The command-line utilities are installed beneath the
\fIProgram\~Files\fR directory; on some
machines this directory is called
\fIProgram\~Files\~(x86)\fR\&.
You can start this utility by running
c:\\Program\~Files\\Regina\\Regina\~8.0\\bin\\sigconvert.exe\&.
.SH "SEE ALSO"
.PP
regina-gui\&.
.SH "AUTHOR"
.PP
This utility was written by Benjamin Burton
<bab@maths.uq.edu.au>\&.
Many people have been involved in the development
of Regina; see the users' handbook for a full list of credits.


utils/sigconvert.cpp

/**************************************************************************
 *                                                                        *
 *  Regina - A Normal Surface Theory Calculator                           *
 *  Convert between different generations of signature                    *
 *                                                                        *
 *  Copyright (c) 2017-2026, Ben Burton                                   *
 *  For further details contact Ben Burton (bab@debian.org).              *
 *                                                                        *
 *  This program is free software; you can redistribute it and/or         *
 *  modify it under the terms of the GNU General Public License as        *
 *  published by the Free Software Foundation; either version 2 of the    *
 *  License, or (at your option) any later version.                       *
 *                                                                        *
 *  As an exception, when this program is distributed through (i) the     *
 *  App Store by Apple Inc.; (ii) the Mac App Store by Apple Inc.; or     *
 *  (iii) Google Play by Google Inc., then that store may impose any      *
 *  digital rights management, device limits and/or redistribution        *
 *  restrictions that are required by its terms of service.               *
 *                                                                        *
 *  This program is distributed in the hope that it will be useful, but   *
 *  WITHOUT ANY WARRANTY; without even the implied warranty of            *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
 *  General Public License for more details.                              *
 *                                                                        *
 *  You should have received a copy of the GNU General Public License     *
 *  along with this program. If not, see <https://www.gnu.org/licenses/>. *
 *                                                                        *
 **************************************************************************/

#include <getopt.h>
#include <cstring>
#include "link/link.h"
#include "triangulation/dim2.h"
#include "triangulation/dim3.h"
#include "triangulation/dim4.h"
#include "triangulation/hidim.h"

static constexpr int maxDim = regina::maxDim();

// Command-line arguments:
int dimension = -1; // 0 indicates knot/link signatures,
                    // -1 indicates not yet specified
int sigGen = 2;

void help() {
    std::cerr <<
R"(Usage: sigconvert [ <isosig> ... | - ]
  -k, --knot                  Input is a knot/link signature
  -2, --dim2                  Input is a 2-D isomorphism signature
  -3, --dim3                  Input is a 3-D isomorphism signature (default)
  -4, --dim4                  Input is a 4-D isomorphism signature
  -d, --dim=<dimension>       Input is an isomorphism signature for the given
                              dimension of triangulation (2 ≤ dimension ≤ )"
        << maxDim << R"()
  -g, --gen=<generation>      Output first-generation signatures (--gen=1) or
                              second-generation signatures (--gen=2, default)
  -v, --version               Show which version of Regina is being used
      --help                  Show this help message
)";
}

bool processLink(const std::string& sig) {
    try {
        auto link = regina::Link::fromSig(sig);
        switch (sigGen) {
            case 1: std::cout << link.knotSig() << std::endl; break;
            case 2: std::cout << link.neoSig() << std::endl; break;
        }
        return true;
    } catch (const regina::InvalidArgument&) {
        std::cerr << "ERROR: Invalid knot/link signature: " << sig
            << std::endl;
        return false;
    }
}

template <int dim> requires (regina::supportedDim(dim))
bool processTri(const std::string& sig) {
    try {
        auto tri = regina::Triangulation<dim>::fromSig(sig);
        switch (sigGen) {
            case 1: std::cout << tri.isoSig() << std::endl; break;
            case 2: std::cout << tri.neoSig() << std::endl; break;
        }
        return true;
    } catch (const regina::InvalidArgument&) {
        std::cerr << "ERROR: Invalid " << dim << "-dimensional "
            "isomorphism signature: " << sig << std::endl;
        return false;
    }
}

int main(int argc, char* argv[]) {
    // Parse the command-line arguments.
    const char* shortOpt = ":d:g:k234v";
    struct option longOpt[] = {
        { "knot", no_argument, nullptr, 'k' },
        { "dim2", no_argument, nullptr, '2' },
        { "dim3", no_argument, nullptr, '3' },
        { "dim4", no_argument, nullptr, '4' },
        { "dim", required_argument, nullptr, 'd' },
        { "gen", required_argument, nullptr, 'g' },
        { "version", no_argument, nullptr, 'v' },
        { "help", no_argument, nullptr, '_' },
        { nullptr, 0, nullptr, 0 }
    };

    int opt;
    char* endptr;
    while ((opt = getopt_long(argc, argv, shortOpt, longOpt, nullptr)) != -1) {
        switch (opt) {
            case 'k':
                if (dimension >= 0 && dimension != 0) {
                    std::cerr << "You cannot pass more than one of "
                        "--knot, --dim2, --dim3, --dim4, or --dim.\n\n";
                    help();
                    return 1;
                }
                dimension = 0;
                break;
            case '2':
                if (dimension >= 0 && dimension != 2) {
                    std::cerr << "You cannot pass more than one of "
                        "--knot, --dim2, --dim3, --dim4, or --dim.\n\n";
                    help();
                    return 1;
                }
                dimension = 2;
                break;
            case '3':
                if (dimension >= 0 && dimension != 3) {
                    std::cerr << "You cannot pass more than one of "
                        "--knot, --dim2, --dim3, --dim4, or --dim.\n\n";
                    help();
                    return 1;
                }
                dimension = 3;
                break;
            case '4':
                if (dimension >= 0 && dimension != 4) {
                    std::cerr << "You cannot pass more than one of "
                        "--knot, --dim2, --dim3, --dim4, or --dim.\n\n";
                    help();
                    return 1;
                }
                dimension = 4;
                break;
            case 'd':
                {
                    int arg = strtol(optarg, &endptr, 10);
                    if (dimension >= 0 && dimension != arg) {
                        std::cerr << "You cannot pass more than one of "
                            "--knot, --dim2, --dim3, --dim4, or --dim.\n\n";
                        help();
                        return 1;
                    }
                    if (*endptr != 0 || arg < 2 || arg > maxDim) {
                        std::cerr << "The dimension must be an integer between "
                            "2 and " << maxDim << " inclusive.\n\n";
                        help();
                        return 1;
                    }
                    dimension = arg;
                }
                break;
            case 'g':
                sigGen = strtol(optarg, &endptr, 10);
                if (*endptr != 0 || ! (sigGen == 1 || sigGen == 2)) {
                    std::cerr << "The signature generation must be 1 or 2.\n\n";
                    help();
                    return 1;
                }
                break;
            case 'v':
                // If other arguments were passed, just silently ignore them
                // for now.  Ideally we would give an error in this scenario.
                std::cout << PACKAGE_BUILD_STRING << std::endl;
                return 0;
            case '_':
                help();
                return 0;
            case '?':
                if (optopt == 0) {
                    // Unknown long option.
                    std::cerr << "Unknown option: " << argv[optind - 1];
                } else {
                    // Unknown short option.
                    std::cerr << "Unknown option: -" <<
                        static_cast<char>(optopt);
                }
                std::cerr << "\n\n";
                help();
                return 1;
            case ':':
                // If an argument is missing, getopt_long() behaves as though
                // the short option were passed, and there is no indication
                // as to whether the user specified a long option instead.
                // For now we will just use the short option in our error
                // message, since the subsequent help text explains both the
                // short and long options anyway.
                std::cerr << "Missing argument: -" << static_cast<char>(optopt);
                std::cerr << "\n\n";
                help();
                return 1;
        }
    }

    // Set any undefined options to their defaults.
    if (dimension < 0)
        dimension = 3;

    if (optind == argc) {
        if (dimension == 0)
            std::cerr << "Please give a knot/link signature.\n\n";
        else
            std::cerr << "Please give an isomorphism signature.\n\n";
        help();
        return 1;
    }

    bool success = true;
    if (optind == argc - 1 && strcmp(argv[optind], "-") == 0) {
        // Read signatures from standard input.
        std::string sig;
        if (dimension == 0) {
            std::getline(std::cin, sig);
            while (! std::cin.eof()) {
                success &= processLink(sig);
                std::getline(std::cin, sig);
            }
        } else {
            std::getline(std::cin, sig);
            while (! std::cin.eof()) {
                regina::select_constexpr<2, maxDim + 1, void>(dimension,
                        [=, &sig, &success](auto dim) {
                    success &= processTri<dim>(sig);
                });
                std::getline(std::cin, sig);
            }
        }
    } else {
        // Read signatures from the command line.
        for (int i = optind; i < argc; ++i)
            if (strcmp(argv[i], "-") == 0) {
                std::cerr <<
R"(A dash (-) indicates that you will supply signatures via standard input.
However, you also appear to be supplying signatures on the command line.
You may use either standard input or the command line, but not both.

)";
                help();
                return 1;
            }

        if (dimension == 0) {
            for (int i = optind; i < argc; ++i)
                success &= processLink(argv[i]);
        } else {
            regina::select_constexpr<2, maxDim + 1, void>(dimension,
                    [=, &success](auto dim) {
                for (int i = optind; i < argc; ++i)
                    success &= processTri<dim>(argv[i]);
            });
        }
    }

    return (success ? 0 : 2);
}

